Rule `no-mixed-spaces-and-tabs` should not apply inside template literals - followup

https://github.com/eslint/eslint/issues/4054 didn't fix all cases.

-

foo.js:

```js
const foo = `${console}
	  foo`;
```

*The second line is `\t  foo`.*

```
~/dev
❯ eslint --version                                                               
1.7.1
                                                                                                                                               
~/dev
❯ eslint foo.js --env=es6 --rule='no-mixed-spaces-and-tabs: 2'

/Users/sindresorhus/dev/foo.js
  2:2  error  Mixed spaces and tabs  no-mixed-spaces-and-tabs

✖ 1 problem (1 error, 0 warnings)
```

-

Can you just make the rule ignore *everything* inside a template literal?

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/27518253-rule-no-mixed-spaces-and-tabs-should-not-apply-inside-template-literals-followup?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

--- a/lib/rules/no-mixed-spaces-and-tabs.js
+++ b/lib/rules/no-mixed-spaces-and-tabs.js
@@ -14,9 +14,7 @@
 module.exports = function(context) {
 
     var smartTabs,
-        templateLocs = [],
-        lastTemplateLocIndex = 0,
-        lastCommentLocIndex = 0;
+        ignoredLocs = [];
 
     switch (context.options[0]) {
         case true: // Support old syntax, maybe add deprecation warning here
@@ -64,7 +62,7 @@ module.exports = function(context) {
     return {
 
         "TemplateElement": function(node) {
-            templateLocs.push(node.loc);
+            ignoredLocs.push(node.loc);
         },
 
         "Program:exit": function(node) {
@@ -77,6 +75,22 @@ module.exports = function(context) {
                 match,
                 lines = context.getSourceLines(),
                 comments = context.getAllComments();
+
+            comments.forEach(function(comment) {
+                ignoredLocs.push(comment.loc);
+            });
+
+            ignoredLocs.sort(function(first, second) {
+                if (beforeLoc(first, second.start.line, second.start.column)) {
+                    return 1;
+                }
+
+                if (beforeLoc(second, first.start.line, second.start.column)) {
+                    return -1;
+                }
+
+                return 0;
+            });
 
             if (smartTabs) {
                 /*
@@ -93,28 +107,18 @@ module.exports = function(context) {
                     var lineNumber = i + 1,
                         column = match.index + 1;
 
-                    for (; lastCommentLocIndex < comments.length; lastCommentLocIndex++) {
-                        if (beforeLoc(comments[lastCommentLocIndex].loc, lineNumber, column)) {
+                    for (var j = 0; j < ignoredLocs.length; j++) {
+                        if (beforeLoc(ignoredLocs[j], lineNumber, column)) {
                             continue;
                         }
-                        if (afterLoc(comments[lastCommentLocIndex].loc, lineNumber, column)) {
-                            break;
+                        if (afterLoc(ignoredLocs[j], lineNumber, column)) {
+                            continue;
                         }
+
                         return;
                     }
 
-                    // make sure this isn't inside of a template element
-                    for (; lastTemplateLocIndex < templateLocs.length; lastTemplateLocIndex++) {
-                        if (beforeLoc(templateLocs[lastTemplateLocIndex], lineNumber, column)) {
-                            continue;
-                        }
-                        if (afterLoc(templateLocs[lastTemplateLocIndex], lineNumber, column)) {
-                            break;
-                        }
-                        return;
-                    }
-
-                    context.report(node, { line: i + 1, column: column }, "Mixed spaces and tabs.");
+                    context.report(node, { line: lineNumber, column: column }, "Mixed spaces and tabs.");
                 }
             });
         }

--- a/tests/lib/rules/no-mixed-spaces-and-tabs.js
+++ b/tests/lib/rules/no-mixed-spaces-and-tabs.js
@@ -29,6 +29,9 @@ ruleTester.run("no-mixed-spaces-and-tabs", rule, {
             code: "\t/*\n\t * Hello\n\t */"
         },
         {
+            code: "// foo\n\t/**\n\t * Hello\n\t */"
+        },
+        {
             code: "/*\n\n \t \n*/"
         },
         {
@@ -37,6 +40,18 @@ ruleTester.run("no-mixed-spaces-and-tabs", rule, {
         },
         {
             code: "/*\n\t */`\n\t   `;",
+            env: { es6: true }
+        },
+        {
+            code: "/*\n\t */var a = `\n\t   `, b = `\n\t   `/*\t \n\t \n*/;",
+            env: { es6: true }
+        },
+        {
+            code: "/*\t `template inside comment` */",
+            env: { es6: true }
+        },
+        {
+            code: "var foo = `\t /* comment inside template\t */`;",
             env: { es6: true }
         },
         {
@@ -49,6 +64,10 @@ ruleTester.run("no-mixed-spaces-and-tabs", rule, {
         },
         {
             code: "`\t   `;",
+            env: { es6: true }
+        },
+        {
+            code: "const foo = `${console}\n\t foo`;",
             env: { es6: true }
         },
         {
@@ -73,6 +92,16 @@ ruleTester.run("no-mixed-spaces-and-tabs", rule, {
                     message: "Mixed spaces and tabs.",
                     type: "Program",
                     line: 2
+                }
+            ]
+        },
+        {
+            code: "\t ;\n/*\n\t * Hello\n\t */",
+            errors: [
+                {
+                    message: "Mixed spaces and tabs.",
+                    type: "Program",
+                    line: 1
                 }
             ]
         },

